docs/decisions/0015-outbox-pattern-and-production-modes.rst

```diff
@@ -0,0 +1,95 @@
+
+- **Ordering**: If multiple events are produced to the same topic, their ordering is preserved. This raises the question of "ordered according to what metric", as concurrency is in play, so the nature of this property may vary by event.
+
+This is only in the general case, as some events may not be connected to database transactions, some consumers might tolerate violations of either atomic success or failure, and not all events may have strict notions of ordering. Hoever, in the general case violations of any of these can result in consistency failures between services that might not be corrected over any time scale.
+
+It's also worth noting a goal we don't have, that of avoiding duplication. At-least-once delivery is acceptible; exactly-once delivery is not required. Double-sends of events are permissible as long as this only happens occasionally (for performance reasons) and does not entail a violation of ordering.
+
+As of 2023-11-09 we produce events in two different ways relative to transactions:
+
+- **Immediate send**: The event is produced to the event bus immediately upon the signal being sent, which will generally occur inside Django's request-level transaction (if using ``ATOMIC_REQUESTS``). This preserves atomicity in the success case as long as the broker is reachable, even if the IDA crashes -- but it does not preserve atomicity when the transaction fails. There is also no ordering guarantee in the case of concurrent requests.
+- **On-commit send**: The event is only sent from a ``django.db.transaction.on_commit`` callback. This preserves atomicity in the failure case, but the IDA could crash after transaction commit but before calling the broker -- or more commonly, the broker could be down or unreachable, and all of the on-commit-produced events would be lost during that interval. Ordering is also not preserved here.
+
+We currently use an ad-hoc mix of immediate and on-commit send in edx-platform, depending on how particular OpenEdxPublicSignals are emitted. For example, the code path for ``COURSE_CATALOG_INFO_CHANGED`` involves an explicit call to ``django.db.transaction.on_commit`` in order to ensure an on-commit send is used. But most signals do not have any such call, and are likely sent immediately. This uncontrolled state reflects our iterative approach to the event bus as well as our choice to start with events that are backed by other synchronization measures which can correct for consistency issues. However, we'd like to start handling events that require stronger reliability guarantees, such as those in the ecommerce space.
```

Comment:
Those other in-process receivers would generally be operating within the request-level transaction though, right? Any which do should have the desired reliability properties already.

Reply:
Not exactly. Any non-trivial in-process event processing would still likely happen async using celery. It just wouldn't require the event bus for communicating across services.

Reply:
Hmm. I guess I'm not sure what you're trying to say.

Reply:
Sorry, maybe I'm getting confused. There are two steps:
1. The call to OpenEdxPublicSignal's send_event, and
2. The call to send an event to a broker in this signal's receiver.

The very first step might be happening immediately, or on-commit, depending on how it was implemented correct? I'm now unclear whether when you are referring to "immediate send" and "on-commit send", whether you are referring to step 1 above, or step 2 above. Also, I'm not clear from an implementation standpoint (without looking more deeply), what combinations of immediate/on-commit are possible across steps 1 and 2.

Once we are on the same page with this, maybe I can make my earlier questions more clear. If this is also not clear, than we can just discuss in person. :)

Reply:
When I refer to immediate or on-commit send, I'm referring to publishing to the broker. It's the only thing that openedx-events has control over -- if the business logic decides to delay calling a signal's send_event until after a transaction has committed, that's out of our control.

(Maybe I should rename those modes to "immediate publish" and "on-commit publish" to distinguish from Django's signal "sending"? Language is hard because we're talking about two layers of pub/sub systems here.)

If the code that sends to a signal does so in transaction.on_commit, the only possible behavior is producing to the broker after transaction completion. If instead the code sends to a signal as soon as possible, then it depends on what mode is configured for that signal -- the event might get published to the broker immediately, or delayed to commit-time.

Reply:
Got it.
1. I think choosing some term like "publish", to differentiate from "send", and clarifying that you will use one with the signal and the other with publishing to the event bus, would help.
2. You can possibly see my confusion due to this sentence:
> We currently use an ad-hoc mix of immediate and on-commit send in edx-platform, depending on how particular OpenEdxPublicSignals are emitted.
Although you possibly mean what you said, it is difficult to understand that you are referring to publishing to the broker in the first half, but referring to the signal send in the second half. Also, the possible combinations were not as clear as your comment above.

Reply:
Yeah, that's... definitely confusing. I'll rework it to use distinct language. :-)

Reply:
Great. And I think adding something of the nature of this comment also helps clarify:

> If the code that sends to a signal does so in transaction.on_commit, the only possible behavior is producing to the broker after transaction completion. If instead the code sends to a signal as soon as possible, then it depends on what mode is configured for that signal -- the event might get published to the broker immediately, or delayed to commit-time.

Although it is somewhat logical/obvious, it also helps differentiate between the two.

Reply:
I think the overall language change I want to make is this:

- "Send" is used only for Django signals
- "Publish" is used for emitting events to the event bus broker

I want to avoid "produce" for the most part, as it leads to awkward phrasing (in my opinion). Produce/consume and publish/subscribe are largely used as synonyms in this space, but publish/subscribe [seems to be more accurate](https://stackoverflow.com/questions/42471870/publish-subscribe-vs-producer-consumer). I still want a way to describe IDAs and requests that _can_ create an event and _try_ to publish it, with the distinction that the publishing may not actually occur in all situations (e.g. in an error situation). For those, "event-producing IDA" and "event-producing request" seem sufficient.

Reply:
Given that there's no current plan to actually implement immediate mode, do you still want that language about immediate and on-commit behaviors?

Reply:
1. Note that we mainly use produce/consume for our docs (e.g. https://openedx.atlassian.net/wiki/spaces/AC/pages/3508699151/How+to+start+using+the+Event+Bus#Producing). I'm not entirely against you using publish/subscribe, but then it feels like you should address this discrepancy. I would wonder how publish/subscribe differs from produce/consume, used everywhere else in the docs.
2. I say you update as you see fit, and then I can answer your question as to whether I think anything is still unclear.